# Transferring Data from Device manufacturer sites to Influxdb
For flexibility and easy of data manipulation, we transfer the data to influx db.
This will also enable us to have the data downloaded and stored in one place - giving us ownership incase our subscription with the device manufacturer expires.

We shall make API calls to the device urls and send the data to the influxdb database. Since different manufacturers have different procedures for API calls, we do upload the data in different batches based on the manufacturer; Quantaq, Clarity, Airgradient

## 0. Set up

### Prepare tools for the project
First install the packages and modules necessary to implement the steps of the project.

The documentation for the influxdb_client_3 library can be found [here](https://docs.influxdata.com/influxdb3/cloud-dedicated/reference/client-libraries/v3/python/)

In [1]:
#Install operational packages to help in 
#organizing and querrying the data 
import pandas as pd
import numpy as np
import requests 
import json
from datetime import datetime
import re                           #Allows to easily remove text using regular expressions
from time import sleep      #Allows pauses during api clalls

#Modules to help interact with influxdb
import os, time
from influxdb_client_3 import InfluxDBClient3, Point

#Modules to modify and manipulate time and dates 
from datetime import date, timedelta

Store important athaurization keys to access different databases

In [2]:
#Load the api key and device serial numbers from the envrionment variable 
%reload_ext dotenv
%dotenv
QUANTAQ_APIKEY = os.environ.get("QUANTAQ_APIKEY")
QUANT001_SN = os.environ.get("QUANT001_SN")
QUANT002_SN = os.environ.get("QUANT002_SN")
CLARITY_APIKEY = os.environ.get("CLARITY_APIKEY")
CLARITY_ORGID = os.environ.get("CLARITY_ORGID")
AIRGRADIENT_APIKEY = os.environ.get("AIRGRADIENT_APIKEY")
EPA_KEY = os.environ.get("EPA_KEY")
EPA_ID = os.environ.get("EPA_ID")
AIRNOW_ID = os.environ.get('AIRNOW_ID')
AIRNOW_KEY = os.environ.get('AIRNOW_KEY')


Create a magic function to help in skipping cells so that they run only once. Some cells especially those that involve uplaoding data to the database, can be skipped so that the data is not duplicated when the cells are are run multiple times

In [3]:
from IPython.core.magic import register_cell_magic

@register_cell_magic
def skip(line, cell):
    return

### Set up influxdb
Prepare the database to receive data from the device manufacturer urls 

In [4]:
#add this to load and upadte the token file
%reload_ext dotenv
%dotenv

#Define some of the arguments required 
#to write the data to the database

token = os.environ.get("INFLUXDB_TOKEN")
org         =   "Air quality - Harrisburg and Philadelphia"
host        =   "https://us-east-1-1.aws.cloud2.influxdata.com"
database    =   "pilot-harrisburg"

client = InfluxDBClient3(host=host, token=token, org=org, database=database)

For automation, Write a function that will later be used to easily upload the data into an influxdb database

In [5]:
def upload_to_inlfuxdb(data_frame_to_save, measurement_name, 
                       index_field, client_name, tag_cols_list):
    """
    A function that takes the name of the measurement(table) 
    and one field name to serve as the index and wites them into
    a defined database on influxdb

    Args:
        data_frame_to_save(df)  : The dataframe to be sent to the database
        measurement_name(string) : The measurement or table name to store the dataset
        index_field(string)     : A timefield from the measurement to be used as an index
                                This also serves as the timeseries reference column
        client_name(object)     : The influxdb object with a defined database 
        tag_cols_list (list)    : A list of column names to act as tags

    Return : 
            Prints that it completed the task
    """ 

    #Make the index field to be the index so that it acts as the timestamp
    data_frame_to_save_index = data_frame_to_save.set_index(index_field)
    
    #Write to the database and the let user know that it is written
    client_name.write(record=data_frame_to_save_index, 
                      data_frame_measurement_name=measurement_name,
                      data_frame_tag_columns=tag_cols_list)
    
    print("Done! Check your influxdb to confirm!")


Since it is not straightforward to delete tables in influxdb; Set the table names here so that when there is a need to delete the table; we can delete the entire database instead and update the table names below

In [6]:
#Measure names defined 
quant_min_table = "min-quantaq-pilot-hbg"
quant_hour_table = "hour-quantaq-pilot-hbg"

clarity_min_table = "min-clarity-pilot-hbg"
clarity_hour_table = "hour-clarity-pilot-hbg"

airgradient_min_table = "min-airgradient-pilot-hbg"
airgradient_hour_table = "hour-airgradient-pilot-hbg"

airnow_hour_table = "airnow-pilot-hbg"


## 1. Quantaq

The guidelines to make API calls to quantaq can be found [here](https://docs.quant-aq.com/software-apis-and-libraries/quantaq-cloud-api). For using python to manipulate data; find the guidance under the [documentation for the py-quantaq](https://quant-aq.github.io/py-quantaq/usage.html#list-all-final-data-for-a-device); which is a quantaq python library for data manipulation. 

### 1.1 Quantaq Setup and Automation

Using the py-QuantAQ library to get data from the web and manipulate it 


In [7]:
%%skip
pip install py-quantaq

In [8]:
#Bring in the the quantaq libray and instantiate the client object
from quantaq.utils import to_dataframe
from quantaq import QuantAQAPIClient

client_quantaq = QuantAQAPIClient(api_key=QUANTAQ_APIKEY)

For automation, write a function that gets the data from quantaq website given a time frame. This function can be used to obtain the intial batch of data and subsequently used to update the database

In [9]:
def get_quantaq_data_devices(devices_sn_list, 
                             start_time, 
                             end_time):
    """
    A function that accepts a list of devices and returns the 
    data of different parameters given a start date and end date
    
    Args:
        devices_sn_list(list) : Sns for devices from which to collect data
        start_time: (string)  : Time in the format of 'yyyy-mm-dd'
        end_time: (string)  : Time in the format of 'yyyy-mm-dd'

    Return: 
        quantaq_all_data_df(Dataframe) : Dataframe with the requested data

    """

    #Initialize the dataframe to be used to make a 
    # singular table to save to the influxdb database
    quantaq_all_data_df = []

    #Collect the data all the devices from commissioning date to jan-13-2026
    for device in devices_sn_list:
        for each in pd.date_range(start = start_time, end = end_time):
            quantaq_all_data_df.append(
                to_dataframe(client_quantaq.data.bydate(sn=device, 
                                                        date=str(each.date())))
            )
    quantaq_all_data_df = pd.concat(quantaq_all_data_df)

    #For simplicity remove the model and weather fields 
    columns_to_remove=[]
    for col in quantaq_all_data_df.columns:
        if col[:5]=='model':
            columns_to_remove.append(col) 
        if col[:3]=='met':
            columns_to_remove.append(col) 
    
    #create a new dataframe without the columns
    quant_all_modified = quantaq_all_data_df.copy()

    quant_all_modified.drop(columns_to_remove, 
                            axis=1, 
                            inplace=True)

    return quant_all_modified

The pandas resampling function uses the timestamp as the index in the sampling process, making it difficult to have two or more devices in the same sampling dataframe since several records are likely to have shared timestamps.

Let's write a function that divides the data into separate dataframes according to the device serial numbers, resamples each and combines the dataframes again

In [10]:
def resample_quantaq(df, frequency='1h'):
    """
    The resample_quantaq takes a quantaq dataframe and 
    Returns a dataframe resampled at a given frequency

    Args:
        df (object)        :A dataframe of quantaq data 
                            collected at minute level
        frequency (string) :Frequency of sampling/aggregating
                            see the resample function in pandas
    
    Return:
        df_all_sampled (object) :A dataframe with data resampled 
                             at the stated interval
    """

    #First convert the timestamp to datetime to pervent any errors
    df['timestamp'] = pd.to_datetime(df['timestamp'])

    #Define the dictionary to be used in resampling so that we can 
    # get the mean of numerical fields wiothout losing categorical ones
    cols_dict = {'co': 'mean', 'no':'mean', 
             'no2':'mean', 'o3':'mean', 
             'pm1':'mean', 'pm10':'mean', ''
             'pm25':'mean', 'raw_data_id':'first', 
             'sn':'first',  
             'timestamp_local':'first', 'url':'first', ''
             'geo.lat':'first', 'geo.lon':'first'}
    
    #Define a dataframe to combine all the dataframes for each device
    df_all_sampled = []    

    for device_sn in pd.unique(df.sn):
        #get a dataframe for each device separately
        df_device = df[df['sn']==device_sn]

        #Resample the dataframe of the device 
        df_device_sampled = (df_device.resample('1h',
                                               on='timestamp',).
                                               agg(cols_dict).
                                               reset_index())

        #Add it to the overall dataframe to be returned 
        df_all_sampled.append(df_device_sampled)

    df_all_sampled = pd.concat(df_all_sampled)    

    return df_all_sampled  
 

### 1.2 QuantAQ Minute Data

##### 1.2.1 First Batch of the QuantAQ minute data to influxDB

Get the first batch of the quantaq data from the quantaq website and store the file as pickle so as to avoid length periods of downloading next time. To allow a buffer in the time ranges; It is advisable to have the start date atleast a day before the intended range and the end date atleast a day after (beyond the range).

In [11]:
%%skip

quantaq_all_data_df = get_quantaq_data_devices(devices_sn_list=['MOD-X-00959','MOD-X-00958'], 
                                               start_time='2026-02-01',
                                               end_time='2026-02-24')

quantaq_all_data_df.to_pickle('data-from-monitors/quantaq_all_devices_first.pkl')

In [12]:
quant_all_first = pd.read_pickle('data-from-monitors/quantaq_all_devices_first.pkl')

Send the initial quantaq batch to the influxdb database for further querrying and vizualization

In [13]:
%%skip
#Use the upload function to send this datframe to inlflux
upload_to_inlfuxdb(data_frame_to_save=quant_all_first, 
                   measurement_name=quant_min_table, 
                   index_field='timestamp', 
                   client_name=client,
                   tag_cols_list=['sn','geo.lat','geo.lon'])

##### 1.3 Update the Quantaq minute data to influxdb Database

Get the most recent date from the intended quantaq influxdb measurement (table). This date is to be used as the start date in the update process. Then use two days from now as the end date - to allow a room for error.

In [14]:
query_date = """SELECT *
                FROM  'min-quantaq-pilot-hbg' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = client.query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from clarity to influxdb
most_recent_quantaq_df = table.to_pandas()
latest_date_quant = str(most_recent_quantaq_df.time[0])[:10]

#Use two days from now as the end date - for the purposes of buffering
end_date_quant = f'{date.today() + timedelta(days=2)}'

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_quant}')
print(f'start: {end_date_quant}')


start: 2026-03-20
start: 2026-03-26


Use the dates above to get the most recent dataframe from quantaq, then uplaod the data to influxdb.

In [15]:
#%%skip
quantaq_update_min = get_quantaq_data_devices(devices_sn_list=['MOD-X-00959','MOD-X-00958'],
                                             start_time=latest_date_quant,
                                             end_time=end_date_quant)



In [16]:
#%%skip
upload_to_inlfuxdb(data_frame_to_save = quantaq_update_min, 
                   measurement_name=quant_min_table, 
                   index_field='timestamp', 
                   client_name=client,
                   tag_cols_list=['sn','geo.lat','geo.lon'])

Done! Check your influxdb to confirm!


### 1.3 QuantAQ Hour Data

##### The First Batch QuantAQ Hour

Resample the data at 1 hour interval before sending it to influxdb. Simple use the batch in the minute category

In [17]:
#%%skip
quant_all_first_hr = resample_quantaq(quant_all_first)

Now upload te resampled data to influxdb

In [18]:
%%skip
#Use the upload function to send this datframe to inlflux
upload_to_inlfuxdb(data_frame_to_save=quant_all_first_hr, 
                   measurement_name=quant_hour_table, 
                   index_field='timestamp', 
                   client_name=client,
                   tag_cols_list=['sn','geo.lat','geo.lon'])

##### 1.3.2 Updating QuantAQ - Hour

Resample the updated data at 1 hour frequency using the minute downloaded data and send it to influxdb

In [19]:
quant_update_hr = resample_quantaq(quantaq_update_min)

#Sometimes the datatypes get mixed up, hence ensure that the 
#data types in this new dataset are consitent with the original one 
quant_update_hr_db = quant_update_hr.astype(quant_all_first_hr.dtypes)

In [20]:
#Use the upload function to send this datframe to inlflux
upload_to_inlfuxdb(data_frame_to_save=quant_update_hr_db, 
                   measurement_name=quant_hour_table, 
                   index_field='timestamp', 
                   client_name=client,
                   tag_cols_list=['sn','geo.lat','geo.lon'])

Done! Check your influxdb to confirm!


## 2. Clarity
The general guidelines for accessing clarity data through the api can be found [here](https://api-guide.clarity.io/); These guideline use examples based on POSTMAN - a webtool for obtaining data using APIs. Clarity has a python library for retrieving data found [here](https://github.com/a2gov/clarityio). Here we use the python library.

### 2.1 Clarity setup and Automation

First install the clarity library

In [21]:
%%skip
pip install clarityio

Make initial connections with the clarity library


In [22]:
import clarityio # Offers methods to connect to the clarity api
from io import StringIO  # Helps to read file as csv and transform to df
api_connection = clarityio.ClarityAPIConnection(api_key=CLARITY_APIKEY, org=CLARITY_ORGID)

Automate the data retrival process from the clarity website

In [23]:
def request_and_fetch_a_report(start_time, 
                               end_time, 
                               frequency_clarity='minute', 
                               metric_select='default'):

    """
    A function that retrieves a one-minute-frequency 
    report in csv-wide format for all devices from clarity 
    A maximum of 30 reports can be retrieved in 24 hours

    Args:
        start_time(string) : Starting time for the period of the report 
                             the format of "yyyy-mm-ddThh:mm:ss.sssZ"
        end_time(string)   : End time for the period of the report in 
                             the format of "yyyy-mm-ddThh:mm:ss.sssZ"
        frequency_clarity(string) : The frequency of aggregation due to resampling
                             example: minute(17 mins), hour, day
        metric_select(string)   : The number of metrics to return
                                  "default" or "all". See clarity documentation
                                  for other options

    Return:
        clarity_report_df(DataFrame) : A dataframe report with the clarity 
                                        data in the specified timeframe
    """

    #Define the base part of the clarity url to be used completed in subsequent requests 
    clarity_api_base_url = 'https://clarity-data-api.clarity.io'

    headers = {"x-api-key": CLARITY_APIKEY}

    # request the report
    result = requests.post(url=clarity_api_base_url + "/v2/report-requests",
                           headers=headers,
                           json={
                               "org": CLARITY_ORGID,
                               "report": "datasource-measurements",
                               "allDatasources": True,
                               "outputFrequency": frequency_clarity,
                               "startTime": start_time,
                               "endTime": end_time,
                               "metricSelect": metric_select
                           })
    result.raise_for_status()
    result_json = result.json()
    reportId = result_json['reportId']

    # poll for its completion
    for i in range(12):
        print("sleeping 15 seconds")
        sleep(15)
        print("fetching report status ... ", end="")
        statusUrl = clarity_api_base_url + f"/v2/report-requests/{reportId}"
        result = requests.get(url=statusUrl, headers=headers)
        result.raise_for_status()
        result_json = result.json()
        print(result_json.get("reportStatus"))
        if result_json.get("reportStatus") != 'in-progress':
            break

    print(result_json)

    #Define a variable to store the csv file
    filename = None

    if result_json.get("reportStatus") == 'succeeded':
        # if it succeeded, fetch the resulting files
        for i, url in enumerate(result_json['urls']):
            with requests.get(url=url, stream=True) as result:
                result.raise_for_status()
                filename = f"extract_{i}.csv"
                # stream to disk
                with open(f"{filename}", "w") as f:
                    for chunk in result.iter_content(1024 * 1024, decode_unicode=True):
                        f.write(chunk)
    
    #Remove all null columns to reduce redudancy
    report_df = pd.read_csv(filename)    
    clarity_report_df = report_df.dropna(axis='columns', how="all")

    #Remove the 'status' columns and those with unconventianal units 
    strings_to_drop = ['Num','status','DwerAqi',
                       'UkDefraAqi', 'PhDenrAqi']
    pattern_clarity = '|'.join(strings_to_drop)
    cols_to_drop = clarity_report_df.filter(regex=pattern_clarity, 
                                            axis=1).columns.to_list()
    clarity_report_less_cols = clarity_report_df.copy()
    clarity_report_less_cols.drop(columns=cols_to_drop, axis=1, inplace=True)

    return clarity_report_less_cols

### 2.2 Clarity Minute Data

##### 2.2.1 The First Batch of the Clarity Minute data

Use the request_and_fetch_a_report function to get the intial batch report from clarity and then uploadd it to inlfuxdb.

In [24]:
%%skip

clarity_first_min = request_and_fetch_a_report(start_time="2026-02-01T00:00:00.000Z",
                                                    end_time="2026-02-24T00:00:00.000Z",
                                                    frequency_clarity='minute')


clarity_first_min.to_pickle('data-from-monitors/clarity_first_min.pkl')

In [25]:

clarity_first_min_db = pd.read_pickle('data-from-monitors/clarity_first_min.pkl')

In [26]:
%%skip
#Upload the report to influxdb
upload_to_inlfuxdb(data_frame_to_save=clarity_first_min_db,
                   measurement_name=clarity_min_table,
                   index_field='time',
                   client_name=client,
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

##### 2.2.2 Updating the Clarity Minute data

To update the clarity database with the latest; first querry for the latest date in the database and use it as the start date to get data from clarity.

In [27]:
query_date = """SELECT *
                FROM  'min-clarity-pilot-hbg' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = client.query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-03-20T00:00:00.000Z
start: 2026-03-26T00:00:00.000Z


Get an updated dataframe from quantaq to upload in the influxdb database

In [28]:
#%%skip
clarity_update_min = request_and_fetch_a_report(start_time=latest_date_clarity,
                                                    end_time=end_date_clarity)


sleeping 15 seconds
fetching report status ... in-progress
sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JBWZDGA69U', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JBWZDGA69U/JBWZDGA69U.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTVJVAS54X&Signature=ebFRMce58c%2Fd%2BN%2BPH7LbN4Wjies%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEM7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQDAZYXgcLecA6GEWGdGN%2B1uVkLkrHk3EMPnuCSFnVJuiQIhAP%2BwUhE%2B%2BaZjkyKSHWXO%2BhOwlmfJSC48%2BELzZh7HszGaKskECJf%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBRoMMDA0NjM5NTEwODg3IgzWYHq2HSJ7HK5nFEMqnQSc3AfePL5l0IsFb3jo6nLmxV3PMvjXo7uv%2F7E7rueMDmuRv4IpiiGob43KznTlMFSGA4iw3QbWF6sKZGY54S0Py1kwgveqgSyB71UGcLAro87Bz3%2BhgAxgQ9s26s%2F4LhV%2B6H75mKYv0jcgrArJmEmdO38%2F5HxxXYMiDUhBx7ZfmgIAwMiVManaBrKjsnSxG52hMl0gXRFU0ZRFpacFtAfwRgkAVdm3Qf0mWqAOzUvX8BCReILS70Y8APK6aP2XANxCNP3u5zvzQmxQ%2B%2BfGKHecsx0WA5gNi

In [29]:
#%%skip
#Upload the report to influxdb
upload_to_inlfuxdb(data_frame_to_save=clarity_update_min,
                   measurement_name=clarity_min_table,
                   index_field='time',
                   client_name=client,
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

Done! Check your influxdb to confirm!


### 2.3 Clarity Hour Data

##### 2.3.1 First Batch of Clarity

Fetch the data from the clarity website; remember to specify the frequency and select the right metrics

In [30]:
%%skip
clarity_first_hour = request_and_fetch_a_report(
    start_time="2026-02-01T00:00:00.000Z",
    end_time="2026-02-24T00:00:00.000Z",
    frequency_clarity="hour",
    metric_select="all",
)


clarity_first_hour.to_pickle("data-from-monitors/clarity_first_hour")

In [31]:

clarity_first_hour_db = pd.read_pickle('data-from-monitors/clarity_first_hour')

In [32]:
%%skip
# Upload the report to influxdb
upload_to_inlfuxdb(
    data_frame_to_save=clarity_first_hour_db,
    measurement_name=clarity_hour_table,
    index_field="startOfPeriod",
    client_name=client,
    tag_cols_list=[
        "datasourceId",
        "sourceId",
        "sourceType",
        "locationLatitude",
        "locationLongitude",
    ],
)

##### 2.3.2 Updating Clarity Hour Data to influxdb

In [33]:
query_date = """SELECT *
                FROM  'hour-clarity-pilot-hbg' 
                ORDER BY time DESC LIMIT 1
             """

# Use the queery on the influxdb database
table = client.query(query=query_date, language="sql")

# Obtain the date to use as a start date in the uplaoding
# the most recent data from clarity to influxdb
most_recent_clarity_df = table.to_pandas()
latest_date_clarity = str(most_recent_clarity_df.time[0])[:10] + "T00:00:00.000Z"

# Use two days from now as the end date - for the purposes of buffering
end_date_clarity = f"{date.today() + timedelta(days=2)}" + "T00:00:00.000Z"

# verify that the date is in the right format and makes sense
print(f"start: {latest_date_clarity}")
print(f"start: {end_date_clarity}")

start: 2026-03-20T00:00:00.000Z
start: 2026-03-26T00:00:00.000Z


In [34]:
clarity_update_df_hour = request_and_fetch_a_report(
    start_time=latest_date_clarity,
    end_time=end_date_clarity,
    frequency_clarity="hour",
    metric_select="all",
)

sleeping 15 seconds
fetching report status ... succeeded
{'reportId': 'JB9YV832IZ', 'reportStatus': 'succeeded', 'message': 'Ready', 'report': 'datasource-measurements', 'urls': ['https://combined-measurements-export-prd.s3.amazonaws.com/historical/JB9YV832IZ/JB9YV832IZ.csv.gz?AWSAccessKeyId=ASIAQCFEJKFTVJVAS54X&Signature=0vMf29yb9JpUs03bbMjL6FUxDfw%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEM7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJIMEYCIQDAZYXgcLecA6GEWGdGN%2B1uVkLkrHk3EMPnuCSFnVJuiQIhAP%2BwUhE%2B%2BaZjkyKSHWXO%2BhOwlmfJSC48%2BELzZh7HszGaKskECJf%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQBRoMMDA0NjM5NTEwODg3IgzWYHq2HSJ7HK5nFEMqnQSc3AfePL5l0IsFb3jo6nLmxV3PMvjXo7uv%2F7E7rueMDmuRv4IpiiGob43KznTlMFSGA4iw3QbWF6sKZGY54S0Py1kwgveqgSyB71UGcLAro87Bz3%2BhgAxgQ9s26s%2F4LhV%2B6H75mKYv0jcgrArJmEmdO38%2F5HxxXYMiDUhBx7ZfmgIAwMiVManaBrKjsnSxG52hMl0gXRFU0ZRFpacFtAfwRgkAVdm3Qf0mWqAOzUvX8BCReILS70Y8APK6aP2XANxCNP3u5zvzQmxQ%2B%2BfGKHecsx0WA5gNiFte6jgDQjpn2JlXTbP3o1lc0%2BYRDnvP%2BfOXMTYZuI8aSo4mTyOtttNDLEYkWC

In [35]:
#Sometimes the datatypes get mixed up, hence ensure that the 
#data types in this new dataset are consitent with the original one 
clarity_update_hour_db = clarity_update_df_hour.astype(clarity_first_hour_db.dtypes)

In [36]:
upload_to_inlfuxdb(data_frame_to_save=clarity_update_hour_db,
                   measurement_name=clarity_hour_table,
                   index_field='startOfPeriod',
                   client_name=client,
                   tag_cols_list=['datasourceId','sourceId','sourceType',
                                  'locationLatitude','locationLongitude']
                   )

Done! Check your influxdb to confirm!


## 3. AirGradient

The general instructions on how to access Airgradient data by API can be found [here](https://www.airgradient.com/professional/cities/undp-toolkit/operating/airgradient-api/). Specific instructions on different options avialable can be found [here](https://api.airgradient.com/public/docs/api/v1/)

### 3.1 AirGradient Automation and setup

Since Airgradient is limitted to a maximum of 10 days in retrieving data, create a function that extend this to more days. 
This is helpful especially in retirving the initial batch of data.

In [37]:
def get_airgradient_data_one(
    location_id,
    start_date,
    end_date,
):
    """
    A function that retrived 5-minute data
    from airgradient in a given time range

    Args:
        location_id : The location id whose data is to be retrieved
        start_date (string)   : A starting date in the format of yyyymmdd
                                Data starts at 00:00:00 utc on this date
        end_date(string)      : An end date in the formart of yyyymmdd
                                Data ends at 00:00:00 utc on this date

    Returns:
        airgradient_df (dataframe) : A dataframe with the airgradient data
    """

    # First convert all the dates to date formart so as to easily perform date operations
    start_date = datetime.strptime(start_date, "%Y%m%d")
    end_date = datetime.strptime(end_date, "%Y%m%d")

    # Initialize the dataframe that will be used to collect other dataframe
    airgradient_df_list = list()

    # Since airgradient allows to retrieve only a maximum of 10 data records
    # We use a loop to segment the data into digestable intervals
    while start_date < end_date:       

        interval_start_date = start_date
        interval_end_date = interval_start_date + timedelta(days=8)

        # Make the new end date to be the starting date for the next cyle
        if interval_end_date > end_date:
            interval_end_date = end_date

        # Convert the dates abve in an airgradient format and
        # use them in the api calls to get the data
        date_from = f"{str(interval_start_date).replace('-','')[:8]}T000000Z"
        date_to = f"{str(interval_end_date).replace('-','')[:8]}T000000Z"

        end_point_params = {
            "location_id": location_id,
            "token": AIRGRADIENT_APIKEY,
            "from": date_from,
            "to": date_to,
        }

        # Define the intial text to be used in the api call
        initial_url = f"https://api.airgradient.com/public/api/v1/locations/{location_id}/measures/past?"
        json_file = requests.get(initial_url, params=end_point_params)

        # Reformat the resulting json file for easy conversion into a df
        json_reformat = json.loads(json_file.content.decode("utf-8"))
        dataframe = pd.json_normalize(json_reformat)

        # Make a list of dataframes to be used to
        # make one single dataframe to be returned
        airgradient_df_list.append(dataframe)

        # Set the start date once again to allow
        # increaments and to break the loop
        start_date = interval_end_date

    # Convert the collected dataframes into one single dataframe
    airgradient_df = pd.concat(airgradient_df_list)

    
    return airgradient_df

Given that we shall be trying to download the data for different locations in a given time frame at once, let's write a function that can take multiple locations.

In [38]:
def get_airgradient_data_many(locations_ids, 
                              start_date,
                                end_date):
    """ 
    Takes a list of location ids and returns a single combined dataframe
    with all the 5-minute interval data collected in a given time range
    """

    #First setup a list to collect the dataframes 
    airgradient_list_many = []

    #collect them in to the list 
    for location in locations_ids:
        airgradient_list_many.append(get_airgradient_data_one(start_date=start_date,
                                                               end_date=end_date,
                                                               location_id=location))

    #Produce one final combined dataframe with all the dataframes 
    airgradient__many_locations_df = pd.concat(airgradient_list_many)  

    #Ensure that the time column is in date format 
    # for easy intergation in influxdb

    airgradient__many_locations_df['timestamp'] = (pd.to_datetime
                                                   (airgradient__many_locations_df
                                                    ['timestamp']))
    
    #Remove any empty rows or columns     
    airgradient__many_locations_df.dropna(axis=1, 
                                          how='all',
                                          inplace=True)
    
    # To avoid, exceeding downloading limits, allow some
    print("sleeping 20 seconds")
    sleep(20)

    return airgradient__many_locations_df                                              



Airgradeint data comes with a number of unwanted fields. This redudancy can sometimes result in compatibility issues especially in updating the database with new data. Create a function that removed some of the unwanted columns 

In [39]:
def clean_airgradient_data(airgradient_raw_df):

    """" 
    Takes an airgradeint dataframe with raw data and 
    ensures that only a few wanted columns are retained 

    """

    #First identify the unawated columns to drop
    cols_to_drop_ag = ['wifi','model', 'tvocIndex', 'noxIndex', 'datapoints']

    airgadient_clean = airgradient_raw_df.drop(columns=cols_to_drop_ag)
    
    return airgadient_clean

Create a resampling function to aggregate the parameters per hour. 
Airgradient does not offer a direct way to pull resampled data for more than 7 days from the resampling date.

In [40]:
def resample_airgradient(df, frequency='1h'):
    
    """
    The resample_airgradient takes a quantaq dataframe and 
    Returns a dataframe resampled at a given frequency

    Args:
        df (object)        :A dataframe of airgradient data 
                            collected at minute level
        frequency (string) :Frequency of sampling/aggregating
                            see the resample function in pandas
    
    Return:
        df_all_sampled (object) :A dataframe with data resampled 
                             at the stated interval
    """

    #First convert the timestamp to datetime to pervent any errors
    #df['timestamp'] = pd.to_datetime(df['timestamp'])

    #Define the dictionary to be used in resampling so that we can 
    # get the mean of numerical fields wiothout losing categorical ones

    cols_dict = {'pm01': 'mean', 'pm02':'mean', 
             'pm02_corrected':'mean', 'pm10_corrected':'mean', 
             'pm10':'mean', 'pm01_corrected':'mean', 
             'pm003Count':'mean', 'atmp':'mean', 'rhum':'mean',
             'rco2':'mean','atmp_corrected':'mean', 'rhum':'mean',
             'rhum_corrected':'mean', 'rco2_corrected':'mean', 
             'locationId':'first',
             'locationName':'first',
             'serialno':'first'             
             }
    
    #Define a dataframe to combine all the dataframes for each device
    df_all_sampled = []    

    for device_sn in pd.unique(df.serialno):
        #get a dataframe for each device separately
        df_device = df[df['serialno']==device_sn]

        #Resample the dataframe of the device 
        df_device_sampled = (df_device.resample('1h',
                                               on='timestamp',).
                                               agg(cols_dict).
                                               reset_index())

        #Add it to the overall dataframe to be returned 
        df_all_sampled.append(df_device_sampled)

    df_all_sampled = pd.concat(df_all_sampled, ignore_index=True)

    #Remove the values without locations 
    df_all_sampled.dropna(axis=0, subset=['locationId'], inplace=True)
     

    return df_all_sampled  

### 3.2 AirGradient Minute Data

Aigradient data is collected at a minimum of 5-minute interval

##### 3.2.1 First Batch of AirGradient Minute

In [41]:
%%skip
# State the locations for which data is to be collected
locations_airgradient = ["175884", "182974", "183437", "183438"]

# Collect the data from the allocations in
# a time of apparoximately a month to date
airgradient_first_min = get_airgradient_data_many(
    start_date="20260201", end_date="20260224", locations_ids=locations_airgradient
)

airgradient_first_min.to_pickle("data-from-monitors/airgradient_first_min.pkl")

Store the data as pickle for easy accessibility and clean it using the clean_airgradient function to ensure that only the useful columns remail

In [42]:
#%%skip
airgradient_min_first_pickle = pd.read_pickle("data-from-monitors/airgradient_first_min.pkl")
airgradient_min_first_db = clean_airgradient_data(airgradient_min_first_pickle)

Upload the data to influxdb

In [43]:
%%skip
#Upload the report to influxdb
upload_to_inlfuxdb(data_frame_to_save=airgradient_min_first_db,
                   measurement_name=airgradient_min_table,
                   index_field='timestamp',
                   client_name=client,
                   tag_cols_list=['locationId','locationName',
                                  'serialno']
                   )

##### 3.2.2 Updating AirGradient Minute

In [44]:
query_date = """SELECT *
                FROM  'min-airgradient-pilot-hbg' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = client.query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from airgradient to influxdb
most_recent_airgradient_df = table.to_pandas()
latest_date_airgradient = str(most_recent_airgradient_df.
                              time[0])[:10].replace('-','')

#Use two days from now as the end date - for the purposes of buffering
end_date_airgradient = f'{date.today() + timedelta(days=2)}'[:10].replace('-','')

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_airgradient }')
print(f'end: {end_date_airgradient}')

start: 20260320
end: 20260326


In [45]:
#Get the updated data 
#State the locations for which data is to be collected
locations_airgradient = ['175884','182974',
                         '183437','183438']

#Use the dates and the locations to get the airgradient data for updates
airgradient_min_update = get_airgradient_data_many(start_date=latest_date_airgradient,
                                                         end_date=end_date_airgradient,
                                                         locations_ids=locations_airgradient)




C:\Users\Owner\AppData\Local\Temp\ipykernel_20816\3358176806.py:19: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  airgradient__many_locations_df = pd.concat(airgradient_list_many)


sleeping 20 seconds


Clean the updates downlaoded and ensure that the data types match those in the first batch

In [46]:
# Clean up and save the updated data
airgradient_min_update_db = clean_airgradient_data(airgradient_min_update)

# Ensure that the data types are matching
airgradient_min_update_db = airgradient_min_update_db.astype(
    airgradient_min_first_db.dtypes,
)

Now send the updates to inlfuxdb

In [47]:
upload_to_inlfuxdb(
    data_frame_to_save=airgradient_min_update_db,
    measurement_name=airgradient_min_table,
    index_field="timestamp",
    client_name=client,
    tag_cols_list=["locationId", "locationName", "serialno"],
)

Done! Check your influxdb to confirm!


### 3.3 Airgradient Hour Data

##### 3.3.1 First Batch AirGardient - Hour

In [48]:
%%skip
# Resample on the minute data to form 1 hour bins
ag_resample_hour = resample_airgradient(airgradient_min_first_db, frequency="hour")

# Ensure that the data types are matching
ag_resample_hour = ag_resample_hour.astype(airgradient_min_first_db.dtypes)

In [49]:
%%skip
# Upload the report to influxdb
upload_to_inlfuxdb(
    data_frame_to_save=ag_resample_hour,
    measurement_name=airgradient_hour_table,
    index_field="timestamp",
    client_name=client,
    tag_cols_list=["locationId", "locationName", "serialno"],
)

##### 3.3.2 Update AirGradient Hour

In [50]:
# Resample on the minute data to form 1 hour bins
ag_resample_hour_update = resample_airgradient(airgradient_min_update_db, frequency="hour")

# Ensure that the data types are matching
ag_resample_hour_update = ag_resample_hour_update.astype(airgradient_min_first_db.dtypes)

In [51]:
# %%skip
# Upload the report to influxdb
upload_to_inlfuxdb(
    data_frame_to_save=ag_resample_hour_update,
    measurement_name=airgradient_hour_table,
    index_field="timestamp",
    client_name=client,
    tag_cols_list=["locationId", "locationName", "serialno"],
)

Done! Check your influxdb to confirm!


## 4. EPA Data

We need API data to give us an average bachmark of what to expect from our models. 
By using data from locations close to our low cost devices, we can get a range of expectation of expectation. 
More guidelines on how to retrieve EPA data using API can be found [here](https://aqs.epa.gov/aqsweb/documents/data_api.html)

### 4.1 EPA Setup and lists
EPA has a set of code lists for states, counties and most parameters. 
It is important to get a summary of these codes.

In [ ]:
%%skip
#First register the the email address to receive a key 
signup_url = 'https://aqs.epa.gov/data/api/signup?email=izabayoalain@gmail.com'
response_json = requests.get(signup_url)
json.loads(response_json.content)

In [ ]:
#get some import lists and tables that will be crucial in retreiving data
states_url = 'https://aqs.epa.gov/data/api/list/states?'
states_params = {'email':EPA_ID,'key':EPA_KEY}
states_json = requests.get(states_url, states_params)
states_epa = pd.json_normalize(json.loads(states_json.content)['Data'])
states_epa.to_pickle("states_codes")


In [13]:
#Pennsylvania Counties codes 
counties_url = 'https://aqs.epa.gov/data/api/list/countiesByState?'
counties_params = {'email':EPA_ID,'key':EPA_KEY, 'state':42}
counties_json = requests.get(counties_url, counties_params)
counties_epa = pd.json_normalize(json.loads(counties_json.content)['Data'])
#counties_epa.to_pickle("counties_codes")

In [14]:
counties_epa

,code,value_represented
0,001,Adams
1,003,Allegheny
2,005,Armstrong
3,007,Beaver
4,009,Bedford
...,...,...
62,125,Washington
63,127,Wayne
64,129,Westmoreland
65,131,Wyoming


Dauphin - 043, Franklin - 055, Lancaster - 071, Cumberland - 041, Adams(Arendtsville) - 001

In [19]:
#%%skip
#Pennsylvania, Duaphin site codes 
sites_url = 'https://aqs.epa.gov/data/api/list/sitesByCounty?'
sites_params = {'email':EPA_ID,'key':EPA_KEY, 'state':42, 'county':'001'}
sites_json = requests.get(sites_url , sites_params)
sites_dauphin_epa = pd.json_normalize(json.loads(sites_json.content)['Data'])
#sites_dauphin_epa.to_pickle("sites_dauphin_epa")

In [20]:
sites_dauphin_epa

,code,value_represented
0,0001,Arendtsville
1,0002,Biglerville
2,8001,None
3,9000,None
4,9991,Arendtsville


From the above, we derive the the site codes for the stations in Harrisburg and Hershey to be; 0401 - Harrisburg, 1100 - Hershey, 0001 - Arendtsville

In [ ]:
%%skip
#Parameters 
class_url = 'https://aqs.epa.gov/data/api/list/classes?'
class_params = {'email':EPA_ID,'key':EPA_KEY}
class_json = requests.get(class_url, class_params)
classes_dauphin_epa = pd.json_normalize(json.loads(class_json.content)['Data'])
classes_dauphin_epa.to_pickle("classes_dauphin_epa")

In [21]:
#%%skip
#Check for all parameter codes 
param_url = 'https://aqs.epa.gov/data/api/list/parametersByClass?'
param_params = {'email':EPA_ID,'key':EPA_KEY, 'pc':'CRITERIA'}
param_json = requests.get(param_url, param_params)
param_dauphin_epa = pd.json_normalize(json.loads(param_json.content)['Data'])
#param_dauphin_epa.to_pickle("param_dauphin_epa")

In [22]:
param_dauphin_epa

,code,value_represented
0,14129,Lead (TSP) LC
1,42101,Carbon monoxide
2,42401,Sulfur dioxide
3,42602,Nitrogen dioxide (NO2)
4,44201,Ozone
5,81102,PM10 Total 0-10um STP
6,85129,Lead PM10 LC FRM/FEM
7,88101,PM2.5 - Local Conditions


From the above we notice the code values of some of the major pollutants we shall monitor. 42101 - Carbon monoxide, 44201	- Ozone, 88101 - PM2.5 - Local Conditions, 42401 - Sulphur dioxide, 42602 - Nitogen dioxide(NO2), 81102 - PM10 Total

Let's get the data now!

In [34]:
sample_url = 'https://aqs.epa.gov/data/api/sampleData/bySite?'
sample_params = {'email':EPA_ID,'key':EPA_KEY, 
                 'param':42401,                  
                 'bdate':20251216, 
                 'edate':20251218,
                 'state':42, 'county':'001', 
                 'site':'0001'}
sample_json = requests.get(sample_url, sample_params)
sample_sites_data_02_2026 = pd.json_normalize(json.loads(sample_json.content)['Data'])
#sample_sites_data_02_2026.to_pickle("sample_sites_data_02_2026")

### 4.2 AirNow Data
Since EPA provides data that is too old - sometimes the latest data maybe 6 months old; we can try to access AirNow data instead. More information about accessing airnow data can be found [here](https://docs.airnowapi.org/)

##### 4.2.1 Automation and Setup

In [52]:
def get_airnow_data(start_date, end_date):
    """
    A function that returns the ozone and pm2.5 data from the
    Harrisburg and Hershey EPA monitors in central PA

    Args:
        start_date(string) : date in the format of yyy-mm-ddThh
        end_date (string)  : date in the format of yyy-mm-ddThh

    Returns:
        harrisburg_air_df (dataframe) : A dataframe with ozone and pm2.5
                                        in Hershey and Harrisburg
    """

    # First define the argumnets to be used in the api call
    url_airnow_1 = f"https://www.airnowapi.org/aq/data/?"
    url_airnow_date = f"startDate={start_date}&endDate={end_date}"
    url_airnow_param = f"&parameters=OZONE,PM25"
    url_airnow_geog = f"&BBOX=-77.378871,39.705655,-76.192348,40.549861"
    url_airnow_format = f"&dataType=B&format=application/json&verbose=1"
    url_airnow_monitor = f"&monitorType=0&includerawconcentrations=0"
    url_airnow_key = f"&API_KEY={AIRNOW_KEY}"

    airnow_url = (
        url_airnow_1
        + url_airnow_date
        + url_airnow_param
        + url_airnow_geog
        + url_airnow_format
        + url_airnow_monitor
        + url_airnow_key
    )
    # print(airnow_url)
    # Use the above to get the jason files
    airnow_json = requests.get(airnow_url).content
    # Ensure that the json file is decoded for easy manipulation
    airnow_json_decoded = airnow_json.decode("utf-8")
    # Obtain a dataframe
    harrisburg_air_df = pd.json_normalize(json.loads(airnow_json_decoded))

    

    return harrisburg_air_df

In [53]:
def clean_airnow_data(airnow_raw_df):
    """ 
    Returns a clean version of the generated dataframe;
    It renames the time field and drops fields that wont
    be helpful during data analysis 

    """
    # Ensure that the timestamp is properly renamed and converted accodingly
    airnow_raw_df["UTC"] = pd.to_datetime(airnow_raw_df["UTC"])
    airnow_raw_time = airnow_raw_df.rename({"UTC": "timestamp"}, axis=1)

    #Drop columns that will most likley not be useful
    cols_to_drop = ['AgencyName']
    airnow_raw_clean = airnow_raw_time.drop(columns= cols_to_drop)

    #Remove missed values which are at times reported as -999.0
    airnow_raw_clean_v1 = airnow_raw_clean[airnow_raw_clean['Value']>0]

    return airnow_raw_clean_v1


##### 4.3.2 First Batch from Airnow

Use the automated functions to get and clean the data. The data should be for nearly all the EPA stations in central PA

In [54]:
%%skip
airnow_first = get_airnow_data(start_date="2026-02-01T00", end_date="2026-02-25T23")
airnow_first_clean = clean_airnow_data(airnow_first)
airnow_first_clean.to_pickle('data-from-monitors/airnow_first_clean.pkl')

In [55]:
airnow_first_clean_db = pd.read_pickle('data-from-monitors/airnow_first_clean.pkl')

Upload the new data in influxdb. Ensure that all the fields that are descriptive are stated under the tags_cols_list

In [56]:
%%skip
# Upload the report to influxdb
upload_to_inlfuxdb(
    data_frame_to_save=airnow_first_clean_db,
    measurement_name=airnow_hour_table,
    index_field="timestamp",
    client_name=client,
    tag_cols_list=[
        "Latitude",
        "Longitude",
        "SiteName",
        "Parameter",
        "FullAQSCode",
        "IntlAQSCode",
        "Unit"
    ]
)

##### 4.2.3 Update Airnow

First obtain the new start and end time

In [57]:
query_date = """SELECT *
                FROM  'airnow-pilot-hbg' 
                ORDER BY time DESC LIMIT 1
             """
    
#Use the queery on the influxdb database
table = client.query(query=query_date, language='sql')

#Obtain the date to use as a start date in the uplaoding 
# the most recent data from clarity to influxdb
most_recent_airnow = table.to_pandas()
latest_date_airnow= str(most_recent_airnow.time[0])[:10] + 'T00'

#Use two days from now as the end date - for the purposes of buffering
end_date_airnow = f'{date.today() + timedelta(days=2)}' + 'T00'

#verify that the date is in the right format and makes sense
print(f'start: {latest_date_airnow}')
print(f'start: {end_date_airnow}')

start: 2026-03-20T00
start: 2026-03-26T00


In [58]:
airnow_update = get_airnow_data(start_date=latest_date_airnow, end_date=end_date_airnow)
airnow_update_db = clean_airnow_data(airnow_update)

In [59]:
# Upload the report to influxdb
upload_to_inlfuxdb(
    data_frame_to_save=airnow_update_db,
    measurement_name=airnow_hour_table,
    index_field="timestamp",
    client_name=client,
    tag_cols_list=[
        "Latitude",
        "Longitude",
        "SiteName",
        "Parameter",
        "FullAQSCode",
        "IntlAQSCode",
        "Unit"
    ]
)

Done! Check your influxdb to confirm!
